nequip/train/schedulefree.py

```diff
@@ -0,0 +1,64 @@
+from .lightning import NequIPLightningModule
+from typing import Dict, Any
+from lightning.pytorch.utilities.exceptions import MisconfigurationException
+from nequip.utils import RankedLogger
+
+logger = RankedLogger(__name__, rank_zero_only=True)
+
+
+class ScheduleFreeLightningModule(NequIPLightningModule):
+    """
+    NequIP LightningModule using Facebook's Schedule-Free optimizer.
+
+    This module wraps the model's optimizer in one of Facebook's Schedule-Free variants.
+    See: https://github.com/facebookresearch/schedule_free
+
+    Args:
+        optimizer (Dict[str, Any]): Dictionary that must include a `_target_`
+            corresponding to one of the Schedule-Free optimizers and other keyword arguments
+            compatible with the Schedule-Free variants.
+    """
+
+    def __init__(self, optimizer: Dict[str, Any], **kwargs):
+        valid_targets = {
+            "AdamWScheduleFree",
+            "SGDScheduleFree",
+            "RAdamScheduleFree",
+        }
+        if "_target_" not in optimizer or not any(
+            optimizer["_target_"].endswith(name) for name in valid_targets
+        ):
+            raise MisconfigurationException(
+                f"Invalid optimizer: expected Schedule-Free optimizer (_target_ ending with one of {valid_targets}), "
+            )
+
+        self.schedulefree_optimizer_class = optimizer["_target_"]
+        super().__init__(optimizer=optimizer, **kwargs)
+
+    @property
+    def evaluation_model(self) -> Any:
+        logger.info("Loading Schedule-Free optimizer weights for evaluation.")
+        return self.model
```

Comment:
The purpose of this method is to return the model that will be used for inference, with the correct weights. For example, see what we're doing in the EMA (exponential moving average) lightning module: https://github.com/mir-group/nequip/blob/60eae375988b07a35b0840eff7db42de076c9a73/nequip/train/ema.py#L94. Is this enough to ensure that the correct weights from the optimizer are loaded in the model. We need to understand how the optimizer handles swapping weights. I think looking at our EMA module would be useful to get a sense of what I mean by swapping weights: https://github.com/mir-group/nequip/blob/main/nequip/train/ema.py. Presumably, the schedule free optimizer is doing that to some degree in analogous ways.

Reply:
Sounds right, worth just printing stuff/inspecting to check.